app/models/manageiq/providers/amazon/agent_coordinator.rb

```diff
@@ -144,68 +131,92 @@ def deploy_agent
       },
-      :subnet_id            => subnets[0].subnet_id,
       :iam_instance_profile => {
         :name => label
       },
+      :network_interfaces   => [{
+        :associate_public_ip_address => true,
+        :delete_on_termination       => true,
+        :device_index                => 0,
+        :subnet_id                   => subnets[0].subnet_id,
+        :groups                      => [security_group_id]
+      }],
       :tag_specifications   => [{
         :resource_type => "instance",
         :tags          => [{
           :key   => "Name",
           :value => label
         }]
       }]
-    )
-    ec2.client.wait_until(:instance_status_ok, :instance_ids => [instance[0].id])
+    ).first
+    ec2.client.wait_until(:instance_status_ok, :instance_ids => [instance.id])
 
     _log.info("Start to load SSA application, this may take a while ...")
 
-    setup_agent(instance[0])
-    _log.info("SSA agent is ready to receive requests.")
+    setup_agent(instance)
+    _log.info("Docker #{agent_docker_name} is loaded. Start to heartbeat.")
 
-    instance[0].id
+    instance.id
   end
 
   def setup_agent(instance)
-    # register if needed
-    register(instance)
+    # Somehow instance.public_dns_name is empty, need to reinitialize to get it back
+    ip = ec2.instance(instance.id).public_dns_name || raise("Failed to get agent's public ip!")
+    key_name = instance.key_name
+    auth_key = get_keypair(key_name).try(:auth_key)
+    _log.error("Key [#{key_name}] is missing. Cannot SSH to the agent:#{instance.id}") if auth_key.nil?
+
+    ssh = LinuxAdmin::SSH.new(ip, agent_ami_login_user, auth_key)
 
     # prepare work directory
-    ssh_commands(instance, ["sudo mkdir -p #{WORK_DIR}; sudo chmod go+w #{WORK_DIR}"], agent_ami_login_user)
+    ssh.perform_commands(["sudo mkdir -p #{WORK_DIR}; sudo chmod go+w #{WORK_DIR}"])
     # scp the default setting yaml file
     create_config_yaml("config.yml")
-    scp_file("config.yml", WORK_DIR.to_s, instance, agent_ami_login_user)
+    scp_file(ip, agent_ami_login_user, auth_key, "config.yml", WORK_DIR)
+    File.delete("config.yml")
+
+    # docker register
+    if agent_ami_registration_required?
+      raise "Need credentials to login" unless docker_auth
+
+      docker_username = docker_auth.userid
+      docker_password = docker_auth.password
+      command_line = "docker login"
+      command_line << " #{agent_docker_name}"
+      command_line << " -u #{docker_username} -p #{docker_password}"
+      ssh.perform_commands([command_line])
+    end
 
     # run docker image
     command_line = "sudo docker run -d --restart=always -v /dev:/host_dev -v #{WORK_DIR}/config.yml:#{WORK_DIR}/config.yml --privileged #{agent_docker_name}"
-    ssh_commands(instance, [command_line.to_s], agent_ami_login_user)
+    ssh.perform_commands([command_line])
   end
 
-  # TODO: for downstream
-  def register(instance)
+  def docker_auth
+    @ems.authentications.find_by(:type => "ssa_docker")
```

Comment:
I thought we were going with `smartstate_docker`  @jerryk55 ?

Reply:
Oh.  That goes to show that I can't remember a name from one moment to the next.  I thought just the label on the Tab was going to be "SmartState Docker" and the authtype could be shortened....  I can change my code and then @hsong-rh can update this.  Apologies all.

Reply:
Yes, we will fix the name in a follow up PR